src/compiler.rs

```diff
@@ -153,4 +227,189 @@ impl Compiler {
 
         rbp.span_offset
     }
+
+    /// Get span of node
+    pub fn get_span(&self, node_id: NodeId) -> Span {
+        *self
+            .spans
+            .get(node_id.0)
+            .expect("internal error: missing span of node")
+    }
+
+    /// Get the source contents of a span of a node
+    pub fn get_span_contents(&self, node_id: NodeId) -> &[u8] {
+        let span = self.get_span(node_id);
+        self.source
+            .get(span.start..span.end)
+            .expect("internal error: missing source of span")
+    }
+
+    // === The following methods are used for the resolution pass ===
+
+    pub fn resolve(&mut self) {
+        if !self.ast_nodes.is_empty() {
+            let last = self.ast_nodes.len() - 1;
+            let last_node_id = NodeId(last);
+
+            self.scope.push(Frame::new(FrameType::Scope, last_node_id));
+            self.scope_stack.push(ScopeId(0));
+
+            self.resolve_node(last_node_id)
+        }
+    }
+
+    pub fn resolve_node(&mut self, node_id: NodeId) {
+        match self.ast_nodes[node_id.0] {
+            AstNode::Block(block_id) => self.resolve_block(node_id, block_id, None),
+            AstNode::Closure { params, block } => {
+                // making sure the closure parameters and body and up in the same scope frame
+                self.enter_scope(block);
+                if let Some(params) = params {
+                    self.resolve_node(params);
+                }
+                let closure_scope = self.exit_scope();
+
+                let AstNode::Block(block_id) = self.ast_nodes[block.0] else {
+                    panic!("internal error: closure's body is not a block");
+                };
+
+                self.resolve_block(block, block_id, Some(closure_scope));
+            }
+            AstNode::Params(ref params) => {
+                // TODO: Remove clone
+                let params = params.clone();
+                for param in params {
+                    self.define_variable(param, false);
+                }
+            }
+            AstNode::Let {
+                variable_name,
+                ty: _,
+                initializer,
+                is_mutable,
+            } => {
+                self.resolve_node(initializer);
+                self.define_variable(variable_name, is_mutable)
+            }
+            AstNode::Variable => self.resolve_variable(node_id),
+            _ => (),
+        }
+    }
+
+    pub fn resolve_block(
+        &mut self,
+        node_id: NodeId,
+        block_id: BlockId,
+        reused_scope: Option<ScopeId>,
+    ) {
+        // TODO: Remove clone
+        let block = self
+            .blocks
+            .get(block_id.0)
+            .expect("internal error: missing block")
+            .clone();
+
+        if let Some(scope_id) = reused_scope {
+            self.enter_existing_scope(scope_id);
```

Comment:
Can you say more about when this happens? Each new scope should have its own variables, for example:

```nushell
if true {
  let x = 3
}
print $x # <-- this errors
```

Reply:
This is to ensure closure parameters end up in the same scope as the closure. Same for custom commands when we implement them.